parser/src/lexer.rs

```diff
@@ -442,23 +443,20 @@ where
 
     fn unicode_literal(&mut self, literal_number: usize) -> Result<char, LexicalError> {
         let mut p: u32 = 0u32;
-        let unicode_error = Err(LexicalError {
+        let unicode_error = LexicalError {
             error: LexicalErrorType::UnicodeError,
             location: self.get_pos(),
-        });
+        };
         for i in 1..=literal_number {
             match self.next_char() {
                 Some(c) => match c.to_digit(16) {
                     Some(d) => p += d << ((literal_number - i) * 4),
-                    None => return unicode_error,
+                    None => return Err(unicode_error),
                 },
-                None => return unicode_error,
+                None => return Err(unicode_error),
             }
         }
-        match wtf8::CodePoint::from_u32(p) {
-            Some(cp) => Ok(cp.to_char_lossy()),
-            None => unicode_error,
-        }
+        std::char::from_u32(p).ok_or(unicode_error)
```

Comment:
This does seem right; cpython keeps characters like `'\uD801'` as `'\uD801'`, while `wtf8` would [replace it with the replacement character](https://docs.rs/wtf8/0.0.3/src/wtf8/lib.rs.html#108-127).

Reply:
Oh no, based on the failing test, it seems CPython strings aren't actually utf-8; they allow surrogate code points (`'\uD800'.encode()` only fails on encode). That'll be questionable for compatibility... in any case, I'd say it would be best to basically embed the surrogate check right here; if `p` is in `D800..=DFFF`, use the REPLACEMENT CHARACTER; otherwise, `std::char::from_u32(p).ok_or(unicode_error)`

Reply:
w/ code

```
        match p {
            '\u{D800}'..='\u{DFFF}' => Ok(std::char::REPLACEMENT_CHARACTER),
            _ =>std::char::from_u32(p).ok_or(unicode_error)
        }
```

I get a bunch of

```
   Compiling rustpython-parser v0.1.1 (/home/dalley/Devel/RustPython/parser)
error: invalid unicode character escape
   --> parser/src/lexer.rs:460:14
    |
460 |             '\u{D800}'..='\u{DFFF}' => Ok(std::char::REPLACEMENT_CHARACTER),
    |              ^^^^^^^^
    |
    = help: unicode escape must not be a surrogate
```

Reply:
Well `p` is a u32, so you'd want to use integer literals instead of character literals.

Reply:
Derp, my bad, I should've looked at the link you pasted.  It should be good now